### DISCUSSION TITLE
Any examples of proper <RootContainer /> implementation? Running into maximum update depth exceed in some cases

### DISCUSSION THREAD

**Discussion by frankcalise** [expo/router#160] [Q&A]
I'm trying to leverage `onStateChanged` / `onReady` as per the `react-navigation` and `expo-router` docs from my top level `_layout.tsx` as that would appear to be where we do the SplashScreen hold up until assets/fonts are loaded, etc.

Passing `initialState` is working without any issues, but when I add onStateChanged for example it creates a loop that runs away.

Moving this config down into an index route didn't seem to change the result. If I drop these two props from `RootContainer` the app functions normally.

Also, the documentation I'm following is here: https://expo.github.io/router/docs/features/container which mentions `you can pass it props like theme and initialState from any screen by using the <RootContainer /> component.` but then later down the page also reads `Avoid using initialState as this is automatically handled by deep links (which Expo Router enables by default).` - this is a bit confusing. However, dropping that prop from the `<RootContainer />` did not solve the depth issue.

Does anyone have any thoughts here or examples?

```tsx
export default function Layout(props: AppProps) {
  const colorScheme = useColorScheme()
  const [isReady, setIsReady] = React.useState(false)

  const navigationRef = RootContainer.useRef()
  const routeNameRef = React.useRef<string>("")

  const {
    initialNavigationState,
    isRestored: isNavigationStateRestored,
  } = useNavigationPersistence(storage, NAVIGATION_PERSISTENCE_KEY)

  const [areFontsLoaded] = useFonts(customFontsToLoad)

  React.useLayoutEffect(() => {
    setTimeout(() => {
      setIsReady(true)
    }, 500)
  }, [])

  if (!areFontsLoaded || !isReady || !isNavigationStateRestored) return <SplashScreen />

  return (
    <>
      <RootContainer
        theme={colorScheme === "dark" ? DarkTheme : DefaultTheme}
        initialState={initialNavigationState}
        onReady={() => {
          console.log('anything here spin away')
          routeNameRef.current = navigationRef.getCurrentRoute().name
        }}
       onStateChange={() => console.log('anything here also spins away')}
      />

      <SafeAreaProvider initialMetrics={initialWindowMetrics}>
        <ErrorBoundary catchErrors={Config.catchErrors}>
          <Stack screenOptions={{ headerShown: false }} />
        </ErrorBoundary>
      </SafeAreaProvider>
    </>
  )
}
```

**Comment by EvanBacon** [comment]
Following up from the [twitter thread](https://twitter.com/Baconbrix/status/1603566499578494980?s=20&t=1kROS8y_RJ9ErRoh59TcKA). The recursive bug should be fixed.

Some other notes:
- Try to rehydrate from the URL directly as this will work better across platforms and require less magic. If users want more control, they should simply move more configuration into the URL query parameters. This could potentially be done by using `Storage.store(useHref().href)` and `useLink().replace(Storage.url)` to rehydrate. There should be multiple ways to do this, if they don't work, it's a bug.
- SafeAreaProvider is added by default. Required for React Navigation, and basically everything.
- React Error Boundaries are [built-in to Expo Router](https://expo.github.io/router/docs/features/errors/#errorboundary) and work per-route. These don't offer much outside of uniformity at the moment but they will have better symbolication, and account for data fetching in a future version. As an example, if you attempt to fetch some info for a route and the server returns an error, you can use the built-in error boundary to display the error and attempt to re-render the page.

**Reply by frankcalise** [reply]
> The recursive bug should be fixed.

Is this going forward? Because I was testing on version 0.0.39.

Appreciate your feedback related to the rehydration, will look to reimplement those patterns as necessary with the new way of thinking. And clean up some of the unnecessary SafeArea/ErrorBoundary stuff.

All that being said the switch has been pretty easy to achieve the same nav structure in [Ignite](https://github.com/infinitered/ignite) (although currently testing it in the [Chain React](https://github.com/infinitered/ChainReactApp2023) but stems from Ignite). I might have you glance at the PR that "converts" over to expo-router if you wouldn't mind after things are cleaned up.

And of course if there is any way I can help you test something specific, hit me up.
